raiden/smart_contracts/TokenNetwork.sol

```diff
@@ -0,0 +1,568 @@
+    {
+        address partner_address;
+
+        partner_address = recoverAddressFromSignature(
+            channel_identifier,
+            nonce,
+            transferred_amount,
+            locksroot,
+            additional_hash,
+            signature
+        );
+
+        uint64 old_nonce = channels[channel_identifier].participants[partner_address].nonce;
+        bytes32 old_locksroot = channels[channel_identifier].participants[partner_address].locksroot;
+
+        // This will reset the transferred amount, invalidating any unlocked locks that were
+        // unlocked but not included in the new locksroot
+        updateParticipantStruct(
+            channel_identifier,
+            partner_address,
+            nonce,
+            locksroot,
+            transferred_amount
+        );
+
+        // Clean up storage for the old (nonce, locksroot)
+        // after the locksroot has been replaced.
+        // This cannot be used anymore to unlock locks anyway.
+        bytes32 old_key = keccak256(old_nonce, old_locksroot);
+        delete channels[channel_identifier].participants[partner_address].withdrawn_locks[old_key];
+
+        TransferUpdated(channel_identifier, msg.sender);
+    }
+
+    /// @notice Unlocks a pending transfer and increases the partner's transferred amount
+    /// with the transfer value. A lock can be unlocked only once per participant.
+    // Anyone can call unlock a transfer on behalf of a channel participant.
+    function unlock(
+        uint channel_identifier,
+        address partner,
+        uint64 expiration_block,
+        uint amount,
+        bytes32 hashlock,
+        bytes merkle_proof,
+        bytes32 secret)
+        stillTimeout(channel_identifier)
+        public
+    {
+        bytes32 key;
+        bytes32 computed_locksroot;
+        bytes memory locked_encoded;
+
+        // Check that the partner is a channel participant.
+        require(channels[channel_identifier].participants[partner].initialized);
+
+        Participant storage partner_state = channels[channel_identifier].participants[partner];
+
+        // An empty locksroot means there are no pending locks
+        require(partner_state.locksroot != 0);
+
+        // The lock must not have expired, it does not matter how far in the
+        // future it would have expired
+        require(expiration_block >= block.number);
+        require(hashlock == keccak256(secret));
+
+        locked_encoded = encodeLock(expiration_block, amount, hashlock);
+        computed_locksroot = computeMerkleRoot(locked_encoded, merkle_proof);
+
+        // Note that unlocked locks have to be re-unlocked after a `transferUpdate` with a
+        // locksroot that does not contain this lock.
+        require(partner_state.locksroot == computed_locksroot);
+
+        // A lock can be withdrawn only once per participant
+        key = keccak256(partner_state.nonce, partner_state.locksroot);
+        require(!partner_state.withdrawn_locks[key]);
```

Comment:
This is incorrect.

The code is allowing only one unlock per merkle root, were we want one unlock per lock in the merkle root.

Reply:
fixed https://github.com/raiden-network/raiden/pull/1286/commits/642fe7432241359ae85408352f18865d4b5f6b6d